File: src/20-challenge/solution.py
from typing import Literal, Final

Gift = Literal["#"]
EmptySlot = Literal["."]
WarehouseCell = Gift | EmptySlot


def fn(
    warehouse: list[list[WarehouseCell]], drops: list[int]
) -> list[list[WarehouseCell]]:
    GIFT: Final[Gift] = "#"
    EMPTY_SLOT: Final[EmptySlot] = "."

    new_warehouse: list[list[WarehouseCell]] = warehouse.copy()

    for drop in drops:
        level = 0
        while level < len(new_warehouse) and new_warehouse[level][drop] == EMPTY_SLOT:
            level += 1

        level -= 1
        if level > -1 and level < len(new_warehouse):
            new_warehouse[level][drop] = GIFT

    return new_warehouse

File: src/20-challenge/solution.ts
type Gift = '#'
type EmptySlot = '.'
type WarehouseCell = Gift | EmptySlot

export default function fn(warehouse: WarehouseCell[][], drops: number[]): WarehouseCell[][] {
	const GIFT: Gift = '#'
	const EMPTY_SLOT: EmptySlot = '.'

	const newWarehouse: WarehouseCell[][] = Array.from(warehouse, (row) => [...row])

	for (const drop of drops) {
		let level = 0
		while (level < newWarehouse.length && newWarehouse[level][drop] === EMPTY_SLOT) level++

		level--
		if (level > -1 && level < newWarehouse.length) newWarehouse[level][drop] = GIFT
	}

	return newWarehouse
}
